rfcs/convergence/native-element-props.md

```diff
@@ -0,0 +1,310 @@
+  </div>
+  {/* optional bookend after slot */}
+</div>
+```
+
+So to specify a value, you'd have to use slot props:
+
+```tsx
+<Input input={{ id: 'foo', value: 'stuff' }} />
+```
+
+The question of where to apply the top-level `ref` has similar considerations and should likely follow the same pattern as native props.
+
+### Principles for any solution
+
+The chosen solution needs to adhere to the following principles (thanks @levithomason):
+
+- User needs to retain access to any part of DOM at any time (can't make a piece of DOM that's unable to receive props)
+- Consider what the user expects when using the component
+
+### For inputs: 3rd-party form validation libraries
+
+To the degree possible, we'd like inputs to work nicely with 3rd-party form libraries, which may have APIs that expect to be able to pass native props to the component and have them applied to the actual `<input>`. (needs more research)
+
+- [Formik](https://github.com/formium/formik)
+- [react-hook-form](https://github.com/react-hook-form/react-hook-form)
+  - Note, for this one they might need to make a wrapper for our component because the library expects to be able to _set_ `.value` on the input, which presents its own challenges. (Someone brought this up for v8 [here](https://github.com/microsoft/fluentui/issues/18126).)
+- [React Final Form](https://github.com/final-form/react-final-form)
+
+## Detailed Design or Proposal
+
+Add a notion of "primary" slot (open to suggestions for the name) where top-level native props and `ref` are applied. This would default to `root` but could be customized (need to work with @bsunderhus to determine implementation approach).
+
+To facilitate passing props to the actual root element when it's not the "primary" slot, explicitly expose the `root` slot in props (possibly with type constraints to prevent passing problematic things). This would be done in all components for consistency.
+
+In the interest of consistency/clarity, customizing the "primary" slot would be discouraged unless a component falls under certain special categories: definitely input components (`Input`, `Checkbox`, etc), and see "Open Issues" below for discussion of other possibilities.
+
+### Example: Checkbox
+
+Suppose `Checkbox` has an `input` slot which has been set as "primary." (Possible APIs for setting this up discussed later.) If the user does this:
+
+```tsx
+<Checkbox name="foo" checked root={{ id: 'bar' }} ref={ref}>
+  sample
+</Checkbox>
+```
+
+Output HTML is roughly like this (a few things omitted for clarity):
+
+```html
+<div id="bar">
+  <!-- ref points to this element -->
+  <input type="checkbox" name="foo" checked />
+  <label>sample</label>
+</div>
+```
+
+If someone really wanted to, they could still pass props explicitly on the `input` slot rather than top-level. This would give the same output HTML as before:
+
+```tsx
+<Checkbox input={{ name: 'foo', checked: true, ref: ref }} root={{ id: 'bar' }}>
+  sample
+</Checkbox>
+```
```

Comment:
IMO we should not do it as it will confuse customers:

```tsx
// ❓ what "name" will have "input" element, "bar" or "foo"?
<Checkbox name="foo" input={{ name: 'foo' }} root={{ id: 'bar' }} />
```

---

There should be clear rules for all components, for example:

```tsx
<Checkbox {...inputProps} root={...rootProps} />
```

- `inputProps` are going to a slot `input`
- `rootProps` are going to a slot `root`
- as there is way to pass props to `input` - component **does not** have `input` slot